tests/test_eigen_tensor.py

```diff
@@ -0,0 +1,256 @@
+import sys
+
+import pytest
+
+np = pytest.importorskip("numpy")
+eigen_tensor = pytest.importorskip("pybind11_tests.eigen_tensor")
+submodules = [eigen_tensor.c_style, eigen_tensor.f_style]
+try:
+    from pybind11_tests import eigen_tensor_avoid_stl_array as avoid
+
+    submodules += [avoid.c_style, avoid.f_style]
+except ImportError:
+    pass
+
+tensor_ref = np.empty((3, 5, 2), dtype=np.int64)
+
+for i in range(tensor_ref.shape[0]):
+    for j in range(tensor_ref.shape[1]):
+        for k in range(tensor_ref.shape[2]):
+            tensor_ref[i, j, k] = i * (5 * 2) + j * 2 + k
+
+indices = (2, 3, 1)
+
+
+def test_import_avoid_stl_array():
+    pytest.importorskip("pybind11_tests.eigen_tensor_avoid_stl_array")
+    assert len(submodules) == 4
+
+
+def assert_equal_tensor_ref(mat, writeable=True, modified=None):
+    assert mat.flags.writeable == writeable
+
+    copy = np.array(tensor_ref)
+    if modified is not None:
+        copy[indices] = modified
+
+    np.testing.assert_array_equal(mat, copy)
+
+
+@pytest.mark.parametrize("m", submodules)
+@pytest.mark.parametrize("member_name", ["member", "member_view"])
+def test_reference_internal(m, member_name):
+
+    if not hasattr(sys, "getrefcount"):
+        pytest.skip("No reference counting")
+    foo = m.CustomExample()
+    counts = sys.getrefcount(foo)
+    mem = getattr(foo, member_name)
+    assert_equal_tensor_ref(mem, writeable=False)
+    new_counts = sys.getrefcount(foo)
+    assert new_counts == counts + 1
+    assert_equal_tensor_ref(mem, writeable=False)
+    del mem
+    assert sys.getrefcount(foo) == counts
+
+
+assert_equal_funcs = [
+    "copy_tensor",
+    "copy_fixed_tensor",
+    "copy_const_tensor",
+    "move_tensor_copy",
+    "move_fixed_tensor_copy",
+    "take_tensor",
+    "take_fixed_tensor",
+    "reference_tensor",
+    "reference_tensor_v2",
+    "reference_fixed_tensor",
+    "reference_view_of_tensor",
+    "reference_view_of_tensor_v3",
+    "reference_view_of_tensor_v5",
+    "reference_view_of_fixed_tensor",
+]
+
+assert_equal_const_funcs = [
+    "reference_view_of_tensor_v2",
+    "reference_view_of_tensor_v4",
+    "reference_view_of_tensor_v6",
+    "reference_const_tensor",
+    "reference_const_tensor_v2",
+]
+
+functions_with_write_flag = [(a, True) for a in assert_equal_funcs] + [
```

Comment:
functools partial might let you clean this up even further. Something like:
```
@pytest.mark.parametrize("func", sum(([
    m.copy_tensor,
    m.copy_fixed_tensor,
    m.copy_const_tensor,
    m.move_tensor_copy,
    m.move_fixed_tensor_copy,
    m.take_tensor,
    m.take_fixed_tensor,
    m.reference_tensor,
    m.reference_tensor_v2,
    m.reference_fixed_tensor,
    m.reference_view_of_tensor,
    partial(m.reference_view_of_tensor_v2, writeable=False),
    m.reference_view_of_tensor_v3,
    partial(m.reference_view_of_tensor_v4, writeable=False),
    m.reference_view_of_tensor_v5,
    partial(m.reference_view_of_tensor_v6, writeable=False),
    m.reference_view_of_fixed_tensor,
    partial(m.reference_const_tensor, writeable=False),
    partial(m.reference_const_tensor_v2, writeable=False),
        ] for m in submodules), []))
``` (untested)

Reply:
Actually, you probably need rpartial or something like so:
```
def rpartial(func, *args):
    return lambda *a: func(*(a + args))
```